pyformex/doc/pyformex.1
.\" Man page generated from reStructeredText.
.
.TH PYFORMEX 1 "2011-12-29" "0.1" "text and X11 processing"
.SH NAME
pyformex \- generate and transform 3D geometry using Python scripts
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.\" TODO: authors and author with name <email>
.
.SH SYNOPSIS
.INDENT 0.0
.INDENT 3.5
.sp
pyformex [options] [ [ file [args] ] ...]
.UNINDENT
.UNINDENT
.SH DESCRIPTION
.sp
This manual page documents briefly the pyformex command.
.sp
pyFormex is a program for generating, transforming and manipulating large geometrical models of 3D structures by sequences of mathematical operations. Thanks to a powerful (Python based) scripting language, pyFormex is very well suited for the automated design of spatial frame structures. It provides a wide range of operations on surface meshes, like STL type triangulated surfaces. There are provisions to import medical scan images. pyFormex can also be used as a pre\- and post\-processor for Finite Element analysis programs. Finally, it might be used just for creating some nice graphics.
.sp
Using pyFormex, the topology of the elements and the final geometrical form can be decoupled. Often, topology is created first and then mapped onto the geometry. Through the scripting language, the user can define any sequence of transformations, built from provided or user defined functions. This way, building parametric models becomes a natural thing.
.sp
While pyFormex is still under development, it already provides a fairly stable scripting language and an OpenGL GUI environment for displaying and manipulating the generated structures.
.SH OPTIONS
.sp
The pyformex command follows the usual GNU command line syntax, with long
options starting with two dashes (\(aq\-\(aq). The non\-option arguments are filenames
of existing pyFormex scripts, possibly followed by arguments, that will be
executed by the pyFormex engine. If a script file uses arguments, it is
responsible for removing the arguments from the list.
.sp
A summary of the most important options is included below.
A complete overview is can be found
from the \(aqpyformex \-h\(aq command. For a complete description of the use and
operation of pyFormex, see the online documentation at
\fI\%http://pyformex.org/doc\fP or the documentation included with the distribution.
.INDENT 0.0
.TP
.B \-\-gui
.
start the GUI (default if no file argument is given)
.TP
.B \-\-nogui
.
do not load the GUI (default if a file argument is given)
.TP
.B \-\-interactive,  \-i
.
Go into interactive mode after processing the command
line parameters. This is implied by the \-\-gui option.
.TP
.BI \-\-config\fB= <file>
.
Read configuration settings from <file>, if it exists.
.TP
.B \-\-nodefaultconfig
.
Skip the default site and user config files. This
option can only be used in conjunction with the
\-\-config option.
.TP
.B \-\-redirect
.
Redirect standard output to the message board (ignored
with \-\-nogui)
.TP
.B \-\-debug
.
display debugging info to standard output
.TP
.B \-\-whereami
.
Show where the pyformex package is installed and exit
.TP
.B \-\-detect
.
Show detected helper software and exit
.TP
.B \-\-version
.
Show the program\(aqs version number and exit.
.TP
.B \-\-help,  \-h
.
Show the help message and exit.
.UNINDENT
.SH SEE ALSO
.sp
\fI\%http://pyformex.org/doc/\fP
.SH AUTHORS
.sp
pyFormex is developed by Benedict Verhegghe with the help of the other
pyFormex project members: Gianluca De Santis, Francesco Iannaccone, Peter Mortier, Tomas Praet, Sofie Van Cauter, Wenxuan Zhou.
.SH AUTHOR
Benedict Verhegghe <benedict.verhegghe@ugent.be>. This manual page was written for the Debian project (and may be used by others).
.SH COPYRIGHT
GPL v3 or higher
.\" Generated by docutils manpage writer.
.\" 
.


History
# $Id$
##
##  This file is part of the pyFormex project.
##  pyFormex is a tool for generating, manipulating and transforming 3D
##  geometrical models by sequences of mathematical operations.
##  Home page: http://pyformex.org
##  Project page:  https://savannah.nongnu.org/projects/pyformex/
##  Copyright (C) Benedict Verhegghe (benedict.verhegghe@ugent.be) 
##  Distributed under the GNU General Public License version 3 or later.
##
##
##  This program is free software: you can redistribute it and/or modify
##  it under the terms of the GNU General Public License as published by
##  the Free Software Foundation, either version 3 of the License, or
##  (at your option) any later version.
##
##  This program is distributed in the hope that it will be useful,
##  but WITHOUT ANY WARRANTY; without even the implied warranty of
##  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##  GNU General Public License for more details.
##
##  You should have received a copy of the GNU General Public License
##  along with this program.  If not, see http://www.gnu.org/licenses/.
##

2009-06 BV: Released pyFormex 0.8

2008-04 BV: Released pyFormex 0.7

2008-02 BV: Started the pyFormex User Meeting

2007-11 BV: Released pyFormex 0.6

2007-08 BV: Released pyFormex 0.5, a major step towards a multifunctional
            program.

2007-02 BV: Released pyFormex 0.4.2, including lots of work on surface
            models.

2006-11-03 BV: Released pyFormex 0.4. This release is marked by the transition
               to the Qt4 widget set.

2006-04-15 BV: Released pyFormex 0.3

2006: pyFormex was used in some of our research projects and proved to be
      a valuable tool. This necessitated and inspired new development work
      on pyFormex. Lots of new features were implemented, too many to be
      listed here.
      Some of the milestones:
      - pyFormex was used in the design of wire stents (see WireStent example).
        The process involved the automatic creation of Finite Element
        Analysis jobs and automatic processing of the results.
      - pyFormex was used to measure geometries obtained from medical images
        and to create a patient specific geometrical hardware model. This
        initiated our further work with surface models.
      - pyFormex was used by civil engineering students for their projects
        in designing shell structures. Some of these will be included as
        examples in future releases. 


2005-12-09 BV : First anniversary release 0.2.2
2005-04-23 BV : Interim 0.2.1 release with lots of new stuff
2005-01-03 BV : First public release on berlios.de (0.2)
2004-12-09 BV : Registered project pyFormex on berlios.de
2004-07-08 BV : Created a first release (0.1.1)
2004-07-02 BV : Created a Subversion repository
2004-06-29 BV :	After reading about Formian, I quickly hacked together
                the Formex class

# End

pyformex.py
#!/usr/bin/env python
# $Id$
##
## This file is part of pyFormex 0.5 Release Mon Jul 30 13:38:48 2007
## pyFormex is a Python implementation of Formex algebra
## Website: http://pyformex.berlios.de/
## Copyright (C) Benedict Verhegghe (benedict.verhegghe@ugent.be) 
##
## This program is distributed under the GNU General Public License
## version 2 or later (see file COPYING for details)
##
"""pyFormex is a program for the design and manipulation of 3D geometry.

"""

import globaldata as GD

import sys
import os.path
from config import Config
from optparse import OptionParser,make_option


###########################  main  ################################


def refLookup(key):
    """Lookup a key in the reference configuration."""
    return GD.refcfg[key]


def printcfg(key):
    try:
        print "!! refcfg[%s] = %s" % (key,GD.refcfg[key])
    except KeyError:
        pass
    print "!! cfg[%s] = %s" % (key,GD.cfg[key])


def gen_pydoc():
    try:
        from pydoc2 import PackageDocumentationGenerator
    except ImportError:
        print "I can not generate the documentation, because you do no have pydoc2."
        return
    excludes = [ 'OpenGL.GL',
                 'OpenGL.GLU',
                 'OpenGL.GLUT',
                 'OpenGL.GLE',
                 'OpenGL.GLX',
                 'numpy',
                 'PyQt4.QtCore',
                 'PyQt4.QtGui',
                 'plugins.properties',
                 ]
    modules = [ 'pyformex', 'config', 'simple', 'utils',
                'gui',
                'plugins',
                ]
    PackageDocumentationGenerator(
        baseModules = modules,
        destinationDirectory = 'doc/html',
        exclusions = excludes,
        ).process ()
    
    

def main(argv=None):
    """This is a fairly generic main() function.

    It is responsible for reading the configuration file(s),
    processing the command line options and starting the application.
    The basic configuration file is 'pyformexrc' located in the pyformex
    directory. It should always be present and be left unchanged.
    You can copy this file to another location if you want to make changes.
    By default, pyformex will try to read the following extra configuration
    files (in this order:
       default settings:     <pyformexdir>/pyformexrc
       system-wide settings: /etc/pyformexrc
       user settings:        $HOME/.pyformexrc
       local settings        $PWD/.pyformexrc
    Also, an extra config file can be specified in the command line.
    Config file settings always override previous ones.
    On exit, the preferences that were changed are written to the last
    read config file. Changed settings are those that differ from the settings
    in all but the last one.
    """
    # this allows us to call main from the interpreter
    if argv is None:
        argv = sys.argv
        
    # get/set the path to the pyformex files, and store it in the config
    pyformexdir = os.path.dirname(os.path.realpath(argv[0]))
    # use a read, not an update, to set the pyformexdir as a variable
    GD.cfg = Config()
    GD.cfg.read("pyformexdir = '%s'\n" % pyformexdir)

    # get/set the user's home dir
    if os.name == 'posix':
        homedir = os.environ['HOME']
    elif os.name == 'nt':
        homedir = os.environ['HOMEDRIVE']+os.environ['HOMEPATH']
    GD.cfg.read("homedir = '%s'\n" % homedir)

    # Process options
    parser = OptionParser(
        usage = "usage: %prog [<options>] [ --  <Qapp-options> ]",
        version = GD.Version,
        option_list=[
        make_option("--gui", help="start the GUI",
                    action="store_true", dest="gui", default=None),
        make_option("--nogui", help="do not load the GUI (default)",
                    action="store_false", dest="gui", default=None),
        make_option("--dri", help="Force the use of Direct Rendering",
                    action="store_true", dest="dri", default=False),
        make_option("--nodri", help="Disables the use of Direct Rendering (overrides the --dri option)",
                    action="store_true", dest="nodri", default=False),
        make_option("--alpha", help="Enable the Alpha Channel buffer",
                    action="store_true", dest="alpha", default=False),
        make_option("--config", help="use file CONFIG for settings",
                    action="store", dest="config", default=None),
         make_option("--nodefaultconfig", help="skip all default locations of config files",
                    action="store_true", dest="nodefaultconfig", default=False),
       make_option("--redirect", help="redirect standard output to the message board (ignored with --nogui)",
                    action="store_true", dest="redirect", default=False),
       make_option("--debug", help="display debugging info to sys.stdout",
                    action="store_true", dest="debug", default=False),
       make_option("--pydoc", help="Generate html documentation from the Python podules",
                    action="store_true", dest="pydoc", default=False),
        ])
    GD.options, args = parser.parse_args()
    GD.print_help = parser.print_help

    GD.debug("Options: %s" % GD.options)

    if GD.options.pydoc:
        gen_pydoc()
        return 0
        
    # Read the config files
    defaults = os.path.join(pyformexdir,"pyformexrc")
    if os.name == 'posix':
        siteprefs = '/etc/pyformexrc'
        prefs = '.pyformexrc'
    elif os.name == 'nt':
        siteprefs = None  # Where does Windoze put site prefs?
        prefs = 'pyformex.cfg'
    homeprefs = os.path.join(homedir,prefs)
    localprefs = os.path.join(os.getcwd(),prefs)

    sysprefs = filter(os.path.exists,[defaults,siteprefs])
    userprefs = filter(os.path.exists,[homeprefs,localprefs])
    if GD.options.nodefaultconfig:
        sysprefs = sysprefs[0:1]
        userprefs = []
    if GD.options.config:
        userprefs.append(GD.options.config)
    refprefs = sysprefs + userprefs[:-1]
    for f in refprefs:
        GD.debug("Reading config file %s" % f)
        GD.cfg.read(f) 
    # Save this config as a reference, then load last user config file
    GD.refcfg = GD.cfg
    GD.cfg = Config(default=refLookup)
    if len(userprefs) > 0:
        GD.preffile = userprefs[-1] # Settings will be saved here
        GD.debug("Reading config file %s" % GD.preffile)
        GD.cfg.read(GD.preffile)
    GD.debug("RefConfig: %s" % GD.refcfg)
    GD.debug("Config: %s" % GD.cfg)

    sys.path[-1:] = GD.cfg.get('syspath',[])

    if GD.options.gui is None and GD.__version__.startswith('0.4') :
        GD.message(GD.Version)
        GD.message("""
!! The pyformex command line options have changed !!
!! By default pyformex will now run without the GUI !!
!! Use pyformex --gui to start the Graphical User Interface.
!! (This message will go away in release 0.5).
""")
    # Run the application with the remaining arguments
    # Importing the gui should be done after the config is set !!
    if GD.options.gui:
        if os.path.isdir(GD.cfg['workdir']):
            # Make the workdir the current dir
            os.chdir(GD.cfg['workdir'])
        else:
            # Save the current dir as workdir
            GD.cfg['workdir'] = os.getcwd()
        from gui.gui import runApp
    else:
        from script import runApp
    
    res = runApp(args)

    #Save the preferences that have changed
    GD.savePreferences()

    # Exit
    return res


#### Go

if __name__ == "__main__":
    sys.exit(main())

#### End
